Mark blocks completed on view in render_xblock view.

symbolist opened the pull request:
Webviews in mobile apps do not encapsulate leaf-blocks in html for vertical blocks. Instead, they use the "chromeless-courseware" `render_xblock` view. This PR adds this missing completion functionality to `render_xblock` view.

Part of [Completion API](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/162247762/Completion+API) work.
Previous PR: https://github.com/edx/edx-platform/pull/16234.

**JIRA tickets**: [OSPR-2441](https://openedx.atlassian.net/browse/OSPR-2441)

**Screenshots**: No change to UI.

**Sandbox URL**: https://pr18273.sandbox.opencraft.hosting/

**Merge deadline**: None

**Testing instructions**:

1. In `/admin/waffle/switch/` add and activate `completion.enable_completion_tracking` to turn on completion tracking.
2. Complete on view tracking happens when the user is visiting courseware pages and when the `render_xblock` view is displayed in the mobile apps.
3. To test courseware pages, visit a course section. When a vertical, which has XBlocks to which this completion strategy is applicable (e.g. html XBlock), is viewed the first time, a `publish_completion` ajax call is made to the server for each child XBlock when it has been viewed from start to end for 5 seconds.
4. To test `render_xblock` view, visit a url of the form `/xblock/<block_id>`. If this completion strategy is applicable to the XBlock, a `publish_completion` ajax call is made to the server when it is viewed for the first time from start to end for 5 seconds.
5. To help with testing, completion data can be reset via the following steps:
```
 ./manage.py lms shell
from completion.models import BlockCompletion
BlockCompletion.objects.all().delete()
```

**Reviewers**
- [x] @pomegranited 
- [x] @staubina or @iloveagent57

edx-webhook commented:
Thanks for the pull request, @symbolist! I've created [OSPR-2441](https://openedx.atlassian.net/browse/OSPR-2441) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

mduboseedx commented:
Hi @symbolist, thank you for the contribution! Feel free to ping me when this is ready for review and passes all the tests.

Thanks!

symbolist commented:
@pomegranited I have added the Python tests. I could not find tests for the courseware views so I have added those as well. There are a few small issues that need to be resolved. But otherwise you can review this now.

symbolist commented:
jenkins run python

symbolist commented:
jenkins run bokchoy

pomegranited commented:
@UmanShahzad  I'm getting this JS error when I load the unit in LMS (same in chromeless view):
```
ReferenceError: markBlocksCompletedOnViewIfNeeded is not defined
```
Are you?

symbolist commented:
@pomegranited Can you pull again? That is due to a wrong attempt at fixing a quality issue that I reverted. https://github.com/edx/edx-platform/pull/18273/files#diff-4a1680b05e9915f0d36cc9fe5f3173b4R26

(Btw you have tagged @UmanShahzad).

pomegranited commented:
Sorry @symbolist , this still isn't working for me 😦   I am seeing the updated code in my browser, so the assets are rebuilding fine, but the `markBlocksCompletedOnViewIfNeeded` method is still undefined.

I couldn't find another example of a requireJS file on the platform that just exports a function in this way -- maybe if you structure it like the other requirejs files and encapsulate the function in an object/class, it will work?

symbolist commented:
@pomegranited In case it is helpful I have refreshed the sandbox with the latest commits and enabled completion on it: https://pr18273.sandbox.opencraft.hosting/.

pomegranited commented on common/lib/xmodule/xmodule/tests/test_vertical.py:
```diff
@@ -136,39 +136,27 @@ def test_render_student_view(self, context):
-        (True, 1.0, 'assertNotIn'),
+        (True, 0.9, True),
+        (False, 0.9, False),
+        (True, 1.0, False),
```
👍  for making these tests better 😄

symbolist commented:
jenkins run all

mduboseedx commented:
@symbolist It looks like tests are passing. Is this ready for review?

bradenmacdonald commented:
@jcdyer FYI, can you take at least a quick look at this too?

jcdyer commented:
@bradenmacdonald Will do.

symbolist commented:
@mduboseedx thanks for keeping track! :) We will let you know as soon as this is ready for review from the edX side (though this is part of a series of work and @jcdyer may already have a reviewer in mind).

symbolist commented:
@jcdyer I have pushed the bokchoy tests to this branch and the tests have passed.

jcdyer commented:
@staubina @iloveagent57 This is ready for an upstream review.  I'm not sure who is planning to take this one.  This is to resolve the issue of completions not getting submitted when viewing pages on mobile.

schenedx commented on lms/djangoapps/courseware/views/views.py:
```diff
@@ -1467,6 +1468,15 @@ def render_xblock(request, usage_key_string, check_if_enrolled=True):
         student_view_context = request.GET.dict()
         student_view_context['show_bookmark_button'] = False
 
+        enable_completion_on_view_service = 'false'
```
Why using string to specify Boolean values?
If the values here are meant for the javascript, can we use Boolean to string `str()`?

symbolist replied:
I have changed this to a Boolean.

schenedx commented on openedx/core/lib/completion.py:
```diff
@@ -0,0 +1,25 @@
+"""Helper methods for completion."""
+
+from xblock.completable import XBlockCompletionMode
```
Should these methods be part of the https://github.com/edx/completion repo?

symbolist replied:
@jcdyer Should these methods be moved onto the `CompletionService`? I see there are a couple of similar helper methods on it.

jcdyer replied:
@symbolist The CompletionService looks like it would be a good place for it.  :+1:

symbolist replied:
Moved to `CompletionService`.

schenedx commented on openedx/core/lib/tests/test_completion.py:
```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+"""
+Tests for completion utils.
```
Same with this test, shouldn't this test be part of https://github.com/edx/completion?

iloveagent57 replied:
edx/completion isn't really the right place, since these tests need modulestore stuff.  I put all of the integration tests between the completion repo and edx-platform here: https://github.com/open-craft/edx-platform/tree/9a6c4033f383f4e05c3390210853ef60d40d4b6a/openedx/tests/completion_integration

symbolist replied:
Some tests were moved to the `completion` package and others to `penedx/tests/completion_integration/test_services.py`.

iloveagent57 commented on common/lib/xmodule/xmodule/vertical_block.py:
```diff
@@ -107,14 +64,24 @@ def student_view(self, context):
+            child_block_context = copy(child_context)
+            if child in child_blocks_to_complete_on_view:
+                child_block_context['wrap_xblock_data'] = {
+                    'mark-completed-on-view-after-delay': completion_service.get_completion_by_viewing_delay_ms()
```
nit: `completion_service.get_completion_by_viewing_delay_ms()` will return the same thing on each call, so you could just store it in a variable outside your `for` loop.

symbolist replied:
I have updated this.

iloveagent57 commented on openedx/core/lib/completion.py:
```diff
@@ -0,0 +1,25 @@
+from xblock.completable import XBlockCompletionMode
+
+
+def should_mark_block_completed_on_view(block):
```
Nit: consider renaming this to something like `can_block_be_completed_on_view`.  Just because a block meets the conditions in the body of this function, doesn't mean it _should_ be marked complete yet - we still have to know if it's already been completed (as you check for below).

symbolist replied:
Renamed to `can_block_be_completed_on_view()`.

jcdyer commented on openedx/core/lib/xblock_utils/__init__.py:
```diff
@@ -92,6 +92,9 @@ def wrap_xblock(
 
+    if context:
+        data.update(context.get('wrap_xblock_data', {}))
+
```
I like the way this lets us put more of the state directly on the completable xblocks without requiring XBlock authors to be explicitly aware of how completion works, and while minimizing the specialness of verticals. :+1:

staubina commented:
Just checking in on this PR, I think @iloveagent57 would be a better reviewer on this ticket as he did work on the original features.  If @iloveagent57 is not available I can provide support.

symbolist commented:
@staubina Thanks! @iloveagent57 is already reviewing it.

schenedx commented:
@symbolist I am available to do another round of review also. Please ping us when you believe this is now good to be reviewed again

symbolist commented:
jenkins run all

iloveagent57 commented:
@symbolist Looks good!  Could you please rebase/squash, then I'll merge this?  Thanks!

edx-webhook commented:
@symbolist 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/18273&entry.1484655318=https://github.com/symbolist&entry.752974735=2018-05-28+13:07&entry.1917517419=2018-06-22+18:46&entry.2133058324=Yes) so we can improve your experience in the future.
